File: src/greedy/MakeBigNumber.java
package greedy;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

//https://programmers.co.kr/learn/courses/30/lessons/42883
public class MakeBigNumber {
    public static String solution(String number, int k) {
        int len = number.length();
        if (len == 1) return number;
        String[] strs = number.split("");
        Set<String> set =
                Stream.of(strs)
                        .sorted((o1, o2) -> Integer.parseInt(o2) - Integer.parseInt(o1))
                        .limit(k)
                        .collect(Collectors.toSet());
        System.out.println(set);

        for (int i = 0; i < strs.length && k > 0; i++) {
            if (!set.contains(strs[i])) {
                strs[i] = "";
                k--;
            }
        }
        return String.join("", strs);
    }

    public static void main(String[] args) {
        String answer = solution("4177252841", 4);
        System.out.println(answer);
    }
}

File: src/greedy/make_big_number.py
# https://programmers.co.kr/learn/courses/30/lessons/42883
from collections import deque
def solution(number, k):
    stack = deque()
    delete_cnt = 0
    for num in number:
        if not stack:
            stack.append(num)
        else:
            while stack and stack[-1] < num and delete_cnt < k:
                stack.pop()
                delete_cnt+=1
            stack.append(num)
    if delete_cnt < k:
        for _ in range(k - delete_cnt):
            stack.pop()
    return ''.join(stack)